lib/charms/layer/apache_bigtop_base.py

```diff
@@ -83,37 +89,68 @@ def render_hiera_yaml(self):
+
+              * namenode
+              * resourcemanager
+              * spark
+              * zk
+              * zk_quorum
+
+            Each master address be applied to all appropriate Hiera properties.
+
+            Note that this is additive.  That is, if the file is rendered again
+            with a different set of hosts, the old hosts will be preserved.
 
-        :param list roles: Optional list of roles this machine will perform.
-            If not given, the ``bigtop_component_list`` layer option will
-            determine which components are configured, and the specific roles
-            will be decided based on whether this machine matches one of the
-            master addresses in the ``hosts`` map.
+        :param list roles: A list of roles this machine will perform, which
+            will be used by `render_site_yaml` to create the configuration for
+            Puppet.
 
-        :param dict overrides: Additional Hiera data to override defaults.
+            If no roles are set, the ``bigtop_component_list`` layer option
+            will determine which components are configured, and the specific
+            roles will be decided based on whether this machine matches one of
+            the master addresses in the `hosts` map.
+
+            Note that this is additive.  That is, if the file is rendered again
+            with a different set of roles, the old roles will be preserved.
+
+        :param dict overrides: A dict of additional data to go in to the
+            ``site.yaml``, which will override data generated from `hosts`
+            and `roles`.
+
+            Note that extra properties set by this are additive.  That is,
+            if the file is rendered again with a different set of overrides,
+            any properties that are not overridden by the new hosts, roles,
+            or overrides will be preserved.
         """
-        bigtop_apt = self.options.get('bigtop_repo-{}'.format(utils.cpu_arch()))
+        hosts = hosts or {}
+        roles = roles or []
+        if isinstance(roles, str):
+            roles = [roles]
+        overrides = overrides or {}
+        site_data = yaml.load(self.site_yaml.text())
 
         # define common defaults
-        site_data = {
-            'bigtop::jdk_package_name': self.options.get('java_package_name'),
+        bigtop_apt = self.options.get('bigtop_repo-{}'.format(utils.cpu_arch()))
+        site_data.update({
+            'bigtop::jdk_preinstalled': True,
             'bigtop::bigtop_repo_uri': bigtop_apt,
             'hadoop::hadoop_storage_dirs': ['/data/1', '/data/2'],
-        }
+        })
 
         # update based on configuration type (roles vs components)
+        roles = set(roles) | set(site_data.get('bigtop::roles', []))
         if roles:
             site_data.update({
                 'bigtop::roles_enabled': True,
-                'bigtop::roles': roles,
+                'bigtop::roles': sorted(roles),
```

Comment:
I am just noting here that we assume that the order in which roles are given has no impact. This seems reasonable, but is it true and will it be true in the future?